### ISSUE TITLE
Improve distributed cluster and API performance: multi-process

### ISSUE THREAD

**Issue by davidjiglesias** [wazuh/wazuh#8251]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.3.0 | Wazuh cluster/API | Manager | All | All |

> **NOTE:** This issue has been updated and the scope and design have changed: https://github.com/wazuh/wazuh/issues/8251#issuecomment-843827385

Hello team,

We should consider using multiple processes to improve the cluster performance. The Wazuh cluster is currently handled by a single process (`wazuh-clusterd`) in each node. Said process is in charge of every single task, including the node server (`master` or `worker`), the **local client** server (attending local petitions for the cluster), the **DAPI** requests queue, the **sendsync** requests queue, the **agent-info** synchronization task and the **integrity** task.

Even though the code is written with asyncio, there is only so much a single process can handle, and the current structure creates a bottleneck that causes huge delays in big and loaded Wazuh environments. Having a single process running and executing all tasks also produces asyncio timeouts once the delays start to take more than 20 seconds (default value), losing the request altogether. 

We can see in the image below the current code structure and the different tasks being handled by `wazuh-clusterd` process:

![image](https://user-images.githubusercontent.com/48716145/115213626-fe64f300-a101-11eb-82b1-bcb0f01316d2.png)

We propose to use multi-processing to handle each one of the different tasks or queues, greatly reducing the total blocking time of the main process (node server), which will be mainly in charge of handling the different tasks distribution and the results, together with the cluster communications.

![image](https://user-images.githubusercontent.com/48716145/115218570-fb203600-a106-11eb-897f-89f13d42add7.png)

In addition, we should make all the necessary changes to the `DAPI` module in order to use [ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#processpoolexecutor) and thus increasing the performance of the API.

Best regards,

David J. Iglesias

**Comment by Selutario** [comment]
## Issue update

Hi team!

I've been doing some research on threading in python and looking to see if it could ultimately make an improvement for our use case in loaded environments. 

With multiple threads we could get each one running on a different core of the processor. However, in the case of Python and a "conventional" implementation like ours, this does not imply that they all run in parallel. In fact, due to interactions with [GIL](https://wiki.python.org/moin/GlobalInterpreterLock) (Global Interpreter Lock), only one of them would be running at a time.

This, which can be good to solve waiting times for I/O, is not useful in our case since we already solved this problem with asyncio.    On the other hand, in cases of heavy CPU usage (our problem with loaded environments), no improvement should be seen. 

It is important to take this information into account before continuing with development, and to investigate further for [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) if necessary. 

Regards,
Selu.

**Comment by davidjiglesias** [comment]
We hold further investigation for design and possible developments until #7366 is completed and the cluster actual state is properly evaluated.

**Comment by vicferpoy** [comment]
## Issue update
Hello team,

As it was stated in https://github.com/wazuh/wazuh/issues/8308#issuecomment-843816160, we are updating this epic issue to use **multi-processing** in both **cluster** and **API**.

The initial schemas are valid for multi-processing as well, so we will keep them. Please keep in mind that when we talk about using multi-processing on the API, we talk about the `DAPI` module.

Regards,
Víctor Fernández

**Comment by davidjiglesias** [comment]
We will finally focus this Epic issue towards making the API able to scale with multiple processes as a configurable option.

**Comment by davidjiglesias** [comment]
API work to be done in https://github.com/wazuh/wazuh/issues/10780. Cluster work already done in v4.3.0
